x-pack/plugins/file_upload/public/importer/geojson_importer/geojson_importer.ts

```diff
@@ -18,161 +17,268 @@ import {
+  private _totalFeatures = 0;
+  private _geometryTypesMap = new Map<string, boolean>();
+  private _invalidFeatures: ImportFailure[] = [];
+  private _prevBatchLastFeature?: Feature;
+  private _geoFieldType: ES_FIELD_TYPES.GEO_POINT | ES_FIELD_TYPES.GEO_SHAPE =
+    ES_FIELD_TYPES.GEO_SHAPE;
+
+  constructor(file: File) {
     super();
+
+    this._file = file;
   }
 
-  public read(data: ArrayBuffer): { success: boolean } {
-    throw new Error('read(data: ArrayBuffer) not supported, use readFile instead.');
+  public destroy() {
+    this._isActive = false;
   }
 
-  protected _createDocs(text: string): CreateDocsResponse {
-    throw new Error('_createDocs not implemented.');
+  public async previewFile(
+    rowLimit?: number,
+    sizeLimit?: number
+  ): Promise<{ features: Feature[]; geoFieldTypes: string[]; previewCoverage: number }> {
+    await this._readUntil(rowLimit, sizeLimit);
+    return {
+      features: [...this._features],
+      previewCoverage: this._hasNext
+        ? Math.round((this._unimportedBytesProcessed / this._file.size) * 100)
+        : 100,
+      geoFieldTypes:
+        this._geometryTypesMap.has('Point') || this._geometryTypesMap.has('MultiPoint')
+          ? [ES_FIELD_TYPES.GEO_POINT, ES_FIELD_TYPES.GEO_SHAPE]
+          : [ES_FIELD_TYPES.GEO_SHAPE],
+    };
   }
 
-  public getDocs() {
-    return this._docArray;
+  public setGeoFieldType(geoFieldType: ES_FIELD_TYPES.GEO_POINT | ES_FIELD_TYPES.GEO_SHAPE) {
+    this._geoFieldType = geoFieldType;
   }
 
-  public setDocs(
-    featureCollection: FeatureCollection,
-    geoFieldType: ES_FIELD_TYPES.GEO_POINT | ES_FIELD_TYPES.GEO_SHAPE
-  ) {
-    this._docArray = [];
-    for (let i = 0; i < featureCollection.features.length; i++) {
-      const feature = featureCollection.features[i];
-      const geometry = feature.geometry as
-        | Point
-        | MultiPoint
-        | LineString
-        | MultiLineString
-        | Polygon
-        | MultiPolygon;
-      const coordinates =
-        geoFieldType === ES_FIELD_TYPES.GEO_SHAPE
-          ? {
-              type: geometry.type.toLowerCase(),
-              coordinates: geometry.coordinates,
-            }
-          : geometry.coordinates;
-      const properties = feature.properties ? feature.properties : {};
-      this._docArray.push({
-        coordinates,
-        ...properties,
-      });
+  public async import(
+    id: string,
+    index: string,
+    pipelineId: string,
+    setImportProgress: (progress: number) => void
+  ): Promise<ImportResults> {
+    if (!id || !index) {
+      return {
+        success: false,
+        error: i18n.translate('xpack.fileUpload.import.noIdOrIndexSuppliedErrorMessage', {
+          defaultMessage: 'no ID or index supplied',
+        }),
+      };
+    }
+
+    let success = true;
+    const failures: ImportFailure[] = [...this._invalidFeatures];
+    let error;
+
+    while (this._features.length > 0 || (this._hasNext && this._isActive)) {
```

Comment:
Shouldn't this condition be `(this._features.length > 0 || this._hasNext) && this._isActive)`?